### ISSUE TITLE
Human-friendly language selection in --format-sort

### ISSUE THREAD

**Issue by CapacitorSet** [yt-dlp/yt-dlp#6755]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

Language selection in --format-sort works, but it references internal numeric identifiers that are not user-friendly (see eg. [this](https://github.com/mpv-player/mpv/issues/8717#issuecomment-1186359552) thread where a user would like --format-sort to prefer the Spanish language, but can't configure yt-dlp to do so).

I think it would be nice to have a human-friendly version to allow users to specify one or more preferred languages. My personal use case for this is watching the bilingual channel [ARTE](https://www.arte.tv/en/), but in general it is useful for anyone consuming resources with multilingual audio.

I think I can implement it relatively easily, but I would like to gather some feedback before delving into it.

### Provide verbose output that clearly demonstrates the problem

- [ ] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [ ] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

**Comment by CapacitorSet** [comment]
Looking closer at the source code for the ARTE extractor, it looks like it [hardcodes](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/arte.py#L162) the selection for the language by looking at the URL (if it contains /en/, it favours, English, and so on) - clever, but entirely opaque to users. My immediate issue is solved, but I feel like it's not the best way to address the problem.

In general, I discovered that the `lang` field is constructed by the extractor with custom rules. While some general criteria make sense (in the case of ARTE, do not prefer the audio description version), the language selection should probably be more general. Especially considering eg. [the linked MPV issue](https://github.com/mpv-player/mpv/issues/8717) that says that yt-dlp by default passes only one audio track to mpv.

**Comment by Grub4K** [comment]
I always felt like streamlining language codes and similar would make a lot of sense. Format filtering is a pain bc of each site possibly using different codes. In my opinion, we should at least try and return fully qualified locales (`de-DE`) and, if not possible, return partial locale (`de`). Sorting or filtering could allow both and convert between them on the fly. Several helper functions should be created for that in `utils`. Extractors could then "opt-in" by just implementing the normalization somewhere in the extraction process, which should hopefully be quite simple. This only becomes problematic if language codes are intransparent and have to be manually mapped and when new languages are added to the site, which would not have that language work with the normalized version. Hopefully this is in line with what @pukkandan expects.

On a side note, mpv hook has several interesting quirks. In my opinion it would make sense to return everything but i dont quite know how to implement that starting from the current version.

**Comment by pukkandan** [comment]
@Grub4K Your request sounds reasonable, but is unrelated to OP?

**Comment by Grub4K** [comment]
> Language selection in --format-sort works, but it references internal numeric identifiers that are not user-friendly

If all extracted language related fields are standardized to an iso locale as described, format sorting, format filtering and possibly even language embedding can be simplified in one go. This will allow users to have a well defined sorting across multiple extractors.

While OP only cared about format sorting, we can get improvements for the others for free with the proposed change. Format sorting (possibly others) just have to be lenient in their processing of locales to allow matching short locales as well.

**Comment by CapacitorSet** [comment]
Nitpick: the "internal numeric identifiers" that I mentioned in the
original issue aren't actual language codes, they're just numbers
arbitrarily defined by the extractor to satisfy some ordering relationship
(eg. in the case of ARTE all languages that do not match the URL have the
digit 0).

So in a way there are two features here - one is exposing the language
field to sorting, and the other is standardizing language codes. Both are
needed, otherwise filtering for en/eng/en-GB/english/English would be quite
unwieldy...

**Comment by pukkandan** [comment]
Yes, that is correct. `lang` is alias for `language_preference`, not for `language`. Technically, `-S language` already works, but not `-S language:fr` since anchor based sorting is not currently implemented for strings.

**Comment by zaednasr** [comment]
@pukkandan is -S "lang:fr" still not functional?

**Comment by DanielSiepmann** [comment]
I'm having the same issue with arte. But it downloads "Französisch__Klare_Sprache" for me, when using `https://www.arte.tv/de/videos/114175-000-A/eisige-welten-naturwunder-im-nordosten-kanadas/` instead of German. That's very frustrating, especially for playlist.

Do I understand it correctly that I need to use `--list-format` for every single video on arte and manually select the id of the language and video and combine them like so:
```
yt-dlp -S 'res:720,ext:mp4' -f VOA-1923+VOA-audio_0-Deutsch --sub-langs de-forced --embed-subs https://www.arte.tv/de/videos/114175-000-A/eisige-welten-naturwunder-im-nordosten-kanadas/
```

Or is there any other option to filter down the preferred language?

**Comment by eirnym** [comment]
I'd like to request an ability to download all audio languages when multiple are available

**Comment by bashonly** [comment]
@eirnym already possible with `--audio-multistreams` and `mergeall`, e.g.:

```
-f "bv+mergeall[vcodec=none]" --audio-multistreams
```

**Comment by eirnym** [comment]
@bashonly 

I need best audio for all languages. `mergeall` would select all formats and download them, which is not what I want to achieve. With `bestaudio` selected `--audio-multistreams`. You can check sample video [here](https://www.youtube.com/watch?v=pEyCkMJ4a0U)

**Comment by eirnym** [comment]
Your filter works for specific video, as codec and bitrate are selected.

Older video or a video from another service would be without audio, as codec could be not opus, or bitrate may be less than 128. 

This issue for human-friendly language selection, and it would be a good idea to add a separate arguments like with subtitles.

**Comment by pukkandan** [comment]
> I'd like to request an ability to download all audio languages when multiple are available

https://github.com/yt-dlp/yt-dlp/issues/1176
